auth/authenticate_with_aws_provider.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snippets
+
+// [START auth_cloud_aws_provider]
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"cloud.google.com/go/auth/credentials/externalaccount"
+	"cloud.google.com/go/storage"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type customAwsProvider struct {
+	awsRegion string
+}
+
+func (acp customAwsProvider) AwsRegion(ctx context.Context, opts *externalaccount.RequestOptions) (string, error) {
+	return acp.awsRegion, nil
+}
+
+func (acp customAwsProvider) AwsSecurityCredentials(ctx context.Context, opts *externalaccount.RequestOptions) (*externalaccount.AwsSecurityCredentials, error) {
+	// Load the AWS default config and retrieve the credentials.
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+	val, err := cfg.Credentials.Retrieve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the AWS credentials into the external account libraries version of AWS security credentials.
+	awsCredentials := &externalaccount.AwsSecurityCredentials{
+		AccessKeyID:     val.AccessKeyID,
+		SecretAccessKey: val.SecretAccessKey,
+		SessionToken:    val.SessionToken,
+	}
+
+	return awsCredentials, nil
+}
+
+// authenticateExplicitWithAdc uses Application Default Credentials
+// to print storage buckets.
+func authenticateExplicitWithAwsProvider(w io.Writer) error {
```

Comment:
suggestion: move the code from the helper functions into this single code snippet function.

Ideally, code-snippets should be "copy/paste runnable" with few (if any) helper functions.

See: https://googlecloudplatform.github.io/samples-style-guide/#introduction